Sync/BatchingClient.swift

```diff
@@ -0,0 +1,218 @@
+            if fitsInBatch(record) {
+                if addToPost(record) {
+                    // Only count the record towards the batch if we could add it to a payload...
+                    addToBatch(record)
+                    return nil
+                } else {
+                    // Otherwise, we don't added the record and just POST away.
+                    return postInBatch(token)
+                }
+            } else {
+                return commitBatch(token)
+            }
+        } else {
+            return addToPost(record) ? nil : start()
+        }
+    }
+
+    private func serialize(record: Record<T>) throws -> UploadRecord {
+        guard let line = self.serializeRecord(record) else {
+            throw SerializeRecordFailure(record: record)
+        }
+
+        let lineSize = line.utf8.count
+        guard lineSize < Sync15StorageClient.maxRecordSizeBytes else {
+            throw RecordTooLargeError(size: lineSize, guid: record.id)
+        }
+
+        return (record.id, line, lineSize)
+    }
+
+    private func addToPost(record: UploadRecord) -> Bool {
+        guard postRecords + 1 <= config.maxPostRecords && postBytes + record.size <= config.maxPostBytes else {
+            return false
+        }
+        postRecords += 1
+        postBytes += record.size
+        records.append(record)
+        return true
+    }
+
+    private func fitsInBatch(record: UploadRecord) -> Bool {
+        return totalRecords + 1 <= config.maxTotalRecords && totalBytes + record.size <= config.maxTotalBytes
+    }
+
+    private func addToBatch(record: UploadRecord) {
+        totalRecords += 1
+        totalBytes += record.size
+    }
+
+    private func postInBatch(token: BatchToken) -> DeferredResponse {
+        // Push up the current payload to the server and reset
+        let lines = self.freezePost()
+        return uploader(lines: lines, ifUnmodifiedSince: self.ifUnmodifiedSince, queryParams: [batchQueryParamWithValue(token)])
+    }
+
+    private func commitBatch(token: BatchToken) -> DeferredResponse {
+        resetBatch()
+        let lines = self.freezePost()
+        let queryParams = [
+            batchQueryParamWithValue(token),
+            NSURLQueryItem(name: "commit", value: "true")
+        ]
+        return uploader(lines: lines, ifUnmodifiedSince: self.ifUnmodifiedSince, queryParams: queryParams)
+            >>== effect(moveForward)
+    }
+
+    private func start() -> DeferredResponse {
+        let postRecordCount = self.postRecords
+        let postBytesCount = self.postBytes
+        let lines = freezePost()
+        return self.uploader(lines: lines, ifUnmodifiedSince: self.ifUnmodifiedSince, queryParams: [batchQueryParamWithValue("true")])
+             >>== effect(moveForward)
+             >>== { response in
+                if let token = response.value.batchToken {
+                    self.batchToken = token
+                    self.totalRecords = postRecordCount
```

Comment:
Nope because since we're starting a batch, the assumption is that these are 0 and we're setting them to the size/count of the start POST call to count them towards the batch. I'll add a comment to clarify.